prophecy-phpunit-autoload.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Prophecy\PhpUnit;
+
+use PHPUnit\Runner\Version as PHPUnit_Version;
+use PHPUnit_Runner_Version;
+
+if (\class_exists('Prophecy\PhpUnit\Autoload', false) === false) {
+
+    /**
+     * Custom autoloader.
+     *
+     * {@internal The code in this file must be PHP cross-version compatible for PHP 5.4 - current!}
+     */
+    class Autoload
```

Comment:
This whole autoload stuff could be replaced by having a `ProphecyTrait.php` file that conditionally requires the right file based on the PHPUnit version (and a fake trait definition in a `if (false)` in that file, to be compatible with the optimized composer autoloader) instead of having a file being loaded eagerly to register the autoloader.

In the case of the Yoast polyfills, I understand the usage of an autoloader given the amount of traits they have. But for a single trait, this is overkill.

Reply:
@stof I hear what you are saying.

If I understand you correctly, you'd like the `files` directive to load the `ProphecyTrait.php`file, which will then contain just an if/else to either load the real trait or declare an empty trait based on the detected PHPUnit version ?

The downside of that is that the (empty) trait will then always be declared as soon as the `vendor/autoload.php` file is included, even when not in a unit test situation.

I suppose I could make the declaration of the empty trait conditional on the PHPUnit version method even being available, but I somehow don't think that's what you meant ?

Reply:
I'd like to keep relying on the Composer autoloader instead of using the `files` section

Reply:
Any preference for the file name of the _real_ trait (as it will have to be renamed) ? Would `ProphecyTraitReal.php` work for you ?

Reply:
Also, as a side-note, which I presume you already considered:
The trade-off of removing the autoloader is that _if_ at some point in the future for PHPUnit 10 or PHP 9, you'd need to introduce a slightly different trait again, the `ProphecyTrait.php` file will just get messier with all the different conditions, but hey, each to their own. I respect your preference as the maintainer of this repo and will make the necessary adjustments.

Reply:
Well, if we need that in the future and we use an autoloader, the same extra logic will still be needed in the autoloader.

Reply:
Just to verify: this would mean changing the `autoload` from `psr-4` (old)/`files` (current) to `classmap` to load the `ProphecyTrait.php` file when needed. Did I interpret that correctly ?

Reply:
well, the `ProphecyTrait.php` file could still be loaded using psr-4 rather than classmap.

and for the 2 non-psr-4 files defining the actual implementations, I suggest naming them `ProphecyTrait.empty.php` and `ProphecyTrait.actual.php`. Using a `.` delimiter rather than a suffix in the file name will ensure that those files are excluded from any PSR-4 loading trying to make them loaded for `Prophecy\PhpUnit\ProphecyTraitEmpty`

Reply:
I've implemented this now as discussed above. It does feel a bit hacky for "abusing" PSR-4 essentially to do the autoloading, but heck, it seems to work.